pkg/reconciler/mongodb/mongodbsource.go

```diff
@@ -14,4 +14,155 @@ See the License for the specific language governing permissions and
+	return reconciler.NewEvent(v1.EventTypeNormal, "MongoDbSourceReconciled", "MongoDbSource reconciled: \"%s/%s\"", namespace, name)
+}
+
+// Reconciler implements controller.Reconciler for MongoDbSource resources.
+type Reconciler struct {
+	// Clients
+	kubeClientSet kubernetes.Interface
+
+	// Lister
+	secretLister corev1listers.SecretLister
+}
+
+// Check that our Reconciler implements Interface
+var _ mongodbsource.Interface = (*Reconciler)(nil)
+
+// Optionally check that our Reconciler implements Finalizer
+//var _ mongodbsource.Finalizer = (*Reconciler)(nil)
+
+// Optionally check that our Reconciler implements ReadOnlyInterface
+// Implement this to observe resources even when we are not the leader.
+//var _ mongodbsource.ReadOnlyInterface = (*Reconciler)(nil)
+
+// Optionally check that our Reconciler implements ReadOnlyFinalizer
+// Implement this to observe tombstoned resources even when we are not
+// the leader (best effort).
+//var _ mongodbsource.ReadOnlyFinalizer = (*Reconciler)(nil)
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, src *v1alpha1.MongoDbSource) reconciler.Event {
+	src.Status.InitializeConditions()
+	src.Status.ObservedGeneration = src.Generation
+
+	// Try to connect to db and see if it works
+	fmt.Printf("----------------------------------------------------------------------------------------------------------------------------------")
+	fmt.Printf(src.Spec.Secret.Name)
+
+	secret, err := r.secretLister.Secrets(src.Namespace).Get(src.Spec.Secret.Name)
+	if err != nil {
+		logging.FromContext(ctx).Error("Unable to read MongoDb credentials secret", zap.Error(err))
+		return err
+	}
+	rawURI, ok := secret.Data["URI"]
+	if !ok {
+		logging.FromContext(ctx).Error("Unable to get MongoDb URI field", zap.Any("secretName", secret.Name), zap.Any("secretNamespace", secret.Namespace))
+		return err
+	}
+	URI := string(rawURI)
+
+	fmt.Printf("raw path: %s", URI)
+	// // Connect to the mongo replica-set
+	// // URI := "mongodb://dev-test:google@10.12.0.34:27017/countries"
+	// URI := fmt.Sprintf("mongodb://%s@10.12.0.34:27017/countries", src.Spec.Secret.Name)
+
+	client, err := mongo.NewClient(options.Client().ApplyURI(URI))
+	if err != nil {
+		logging.FromContext(ctx).Error("Couldn't connect to DB: %w", err)
+		return err
+	}
+	err = client.Connect(ctx)
+	if err != nil {
+		logging.FromContext(ctx).Error("Couldn't connect to DB: %w", err)
+		return err
+	}
+	// Make sure it is connected
+	err = client.Ping(ctx, readpref.Primary())
+	if err != nil {
+		logging.FromContext(ctx).Error("Couldn't connect to DB: %w", err)
+		return err
+	}
+	defer client.Disconnect(ctx)
+
+	// See if database exists
+	if client.Database(src.Spec.Database) == nil {
+		logging.FromContext(ctx).Error("Couldn't find DB %s: %w", src.Spec.Database, err)
+	}
+
+	// See if collection exists
+	if client.Database(src.Spec.Database).Collection(src.Spec.Collection) == nil {
```

Comment:
I think here you need to check whether src.Spec.Collection != "", and then do this. Because our collection was optional.